### PR TITLE
MDP Playground integration into bsuite

### PR DESCRIPTION
**Pull request by RaghuSpaceRajan**
Hi @iosband and @yotam,

Hope you're doing well! 

Following our discussions, I added MDP Playground experiments into bsuite for the following dimensions of MDP Playground: **Delay**, **Transition Noise**, **Reward Noise**, **Reward Sparsity**, **Rewardable Sequence Length**

#### Here is a short summary of the changes we have made:

* Added MDP Playground environments and experiments into bsuite
  Updated analysis Jupyter notebook:
   * Added a spoke in the bsuite spider plot for MDP Playground environments.
   * Added additional analyses cells for individual MDP Playground experiments:
     * Delay, Transition Noise, Reward Noise, Reward Sparsity, Rewardable Sequence Length

* Added a guide on how to add new experiments into bsuite to `CONTRIBUTING.md`

* Updated setup.py to include `mdp-playground` as a dependency

* Changed HPs for `A2C` because performance was very noisy with the old ones

* Removed Jupyter notebook conflicts with `deepmind:master`

* Removed  `.gitignore` conflicts with `deepmind:master`

#### Improvements still needed:
* The code still needs to conform to the Python style that bsuite follows.
* The tests probably need some improvement.

Please let us know your inputs and feedback on what to do next!

Best regards,
Raghu Rajan.

Co-authored-by: suresh-guttikonda <guttikondasuresh7@gmail.com>
Co-authored-by: guttikon <guttikon@informatik.uni-freiburg.de>

### CODE REVIEW

**Comment by suresh-guttikonda** [comment]
@googlebot I fixed it.

**Comment by suresh-guttikonda** [comment]
@googlebot I fixed it.

**Comment by RaghuSpaceRajan** [comment]
Hi @iosband and @yotam,

Could you please restart the workflow tests? I managed to fix the failing _pytype_ tests in the latest commit and _pytest_ tests are also passing locally.

File: bsuite/baselines/tf/actor_critic/run.py [hunk 1]
```diff
@@ -43,8 +43,8 @@
 # algorithm
 flags.DEFINE_integer('seed', 42, 'seed for random number generation')
 flags.DEFINE_integer('num_hidden_layers', 2, 'number of hidden layers')
-flags.DEFINE_integer('num_units', 64, 'number of units per hidden layer')
-flags.DEFINE_float('learning_rate', 1e-2, 'the learning rate')
+flags.DEFINE_integer('num_units', 50, 'number of units per hidden layer')
```
Comment (yotam) [inline bsuite/baselines/tf/actor_critic/run.py:46]: We should do any agent changes in a separate commit

File: bsuite/environments/mdp_playground.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# pylint: disable=g-bad-file-header
+# Copyright 2019 .... All Rights Reserved.
```
Comment (yotam) [inline bsuite/environments/mdp_playground.py:2]: 2021 :)

File: bsuite/environments/mdp_playground.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""The MDP Playground reinforcement learning environment."""
+
+# Import all packages
+
+from mdp_playground.envs import RLToyEnv #mdp_playground
+
+# import collections
```
Comment (yotam) [inline bsuite/environments/mdp_playground.py:22]: Please remove the commented-out import

File: bsuite/environments/mdp_playground.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+from dm_env import specs
+from dm_env import StepType
+import gym
+import numpy as np
+from typing import Any
+
+# def ohe_observation(obs):
+
+class DM_RLToyEnv(base.Environment):
+  """A wrapper to convert an RLToyEnv Gym environment from MDP Playground to a
```
Comment (yotam) [inline bsuite/environments/mdp_playground.py:36]: It's sufficient to write "A dm_env wrapper for the Gym RLToyEnv." here.

File: bsuite/environments/mdp_playground.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    self.max_episode_len = max_episode_len
+    self._raw_return = 0.
+    self._best_episode = 0.
+    self._episode_return = 0.
+
+    self.bsuite_num_episodes = sweep.NUM_EPISODES
+
+    super(DM_RLToyEnv, self).__init__()
+    # Convert gym action and observation spaces to dm_env specs.
```
Comment (yotam) [inline bsuite/environments/mdp_playground.py:52]: There's some more commented-out code here

File: bsuite/environments/mdp_playground.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+from typing import Any
+
+# def ohe_observation(obs):
+
+class DM_RLToyEnv(base.Environment):
+  """A wrapper to convert an RLToyEnv Gym environment from MDP Playground to a
+  base.Environment which is a subclass of dm_env.Environment.
+  Based on the DMEnvFromGym in gym_wrapper.py"""
+
+  def __init__(self, max_episode_len: int = 100, **config: Any):
```
Comment (yotam) [inline bsuite/environments/mdp_playground.py:40]: Could you rename `config` to something like `gym_make_kwargs`? Ideally the type would be something like `Mapping[str, Any]` too.

File: bsuite/experiments/mdp_playground/analysis.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+GOOD_EPISODE = 50
+TAGS = sweep.TAGS
+
+def score(df: pd.DataFrame) -> float:
+  """Output a score for MDP Playground."""
+  df = mdpp_preprocess(df_in=df)
+  regret_score = plotting.ave_regret_score(
+      df, baseline_regret=BASE_REGRET, episode=NUM_EPISODES)
+
+  norm_score = 1.0 * regret_score # 2.5 was heuristically chosen value to get Sonnet DQN to score approx. 0.75, so that better algorithms like Rainbow can get score close to 1. With a bigger NN this would mean an unclipped score of 1.1 for Sonnet DQN, which is fair I think. However, a2c_rnn even reached 2.0 on this scale. DQN may be not performing as well because its epsilon is not annealed to 0.
```
Comment (yotam) [inline bsuite/experiments/mdp_playground/analysis.py:39]: I don't understand the comment I'm afraid. Was the value here 2.5 at some other point>

File: bsuite/experiments/mdp_playground/analysis.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  print("unclipped score:", norm_score)
+  norm_score = np.clip(norm_score, 0, 1)
```
Comment (yotam) [inline bsuite/experiments/mdp_playground/analysis.py:41]: nit: Prefer `return np.clip(...)` rather than creating a variable and returning it on the next line.

**Review by yotam — COMMENTED** [review]
Hi Raghu, I've read through some of the files and left a few initial comments, just on (Google) Python style and so on. I'm unsure about adding to the "main" bsuite.py from the perspective of expanding dependencies (e.g. the gym dependency did not exist before https://github.com/deepmind/bsuite#using-bsuite-in-openai-gym-format).

Let's go through Ian's thoughts via email first about the experiments themselves, then figure out where to go with this proposal.
